### PR TITLE
Unify Demonstration Learning Algorithms

### LINKED ISSUE #246: Common reward model interface
Currently our `RewardNet` class is designed for `AIRL`. In particular, having separate `train` and `test` outputs is specific to AIRL, and does not make any sense for most others algorithms.

We currently have an alternative interface `tabular_irl.RewardModel`. This is intended for JAX -- but it'd be incompatible with `RewardNet` even after being ported to PyTorch. (And note that it's not actually specific to tabular methods -- although some of the concrete subclasses are.)

In the future we might add other algorithms like GCL, DRLHP, etc which also use rewards but do not have this train vs test distinction.

My proposal would be to make an abstract `RewardNet` or `RewardModel` class (either name seems fine) that is minimal and so should work across all algorithms. In particular, I imagine the `__init__` would take in an `observation_space` and `action_space`, there would be a single abstract method `reward` (v.s. `reward_train` and `reward_test` now) and a concrete implementation `predict` taking NumPy arrays. (You might also need some other things for PyTorch, e.g. `device`, that we have now.)

We can then have concrete instances for various different reward models. I would suggest having one that is an MLP reward model (with boolean flags indicating whether it uses observation, action, next observation and dones as input), one that is an MLP potential shaping only, a class that lets you take linear combinations of reward models, and then a shaped MLP can be defined in terms of these.

For AIRL you could then feed in an MLP reward model and an MLP potential shaping, it can internally sum them and have the train vs test distinction -- with no need for a separate class for this.

Note I've done something similar in https://github.com/HumanCompatibleAI/evaluating-rewards/blob/master/src/evaluating_rewards/rewards/base.py which has been working well (and I iterated on a couple of designs before getting to this stage). But that's all TF code, so cannot be used verbatim.

### PR DESCRIPTION
**Pull request by AdamGleave**
Introduces a new ABC `DemonstrationAlgorithm` that inherits from `BaseImitationAlgorithm`. This interface is fairly minimal: it takes a `demonstrations` constructor argument has a `set_demonstrations` method, both taking a variety of possible demonstration types (sequence of trajectories, transitions, or batches of transitions).

Modifies `AdversarialTrainer` (and so `GAIL` and `AIRL`), `BC`, and `MCEIRL` (new class, previously a function) to inherit from `DemonstrationAlgorithm`.

While the ABC itself doesn't specify much, I've also started trying to make the APIs at least feel more similar in other ways (similar parameter and method naming conventions). The algorithms just are different in a lot of ways, so I don't think making them have an identical API is going to work. 

In a future PR I plan to start unifying the training scripts, and so I'll see from that what the pain points / similarities are, and may make further refinements.

Also makes a variety of minor code cleanups. In particular I ran `darglint` (docstring linting tool) across all the imitation algorithms and fixed errors it flagged.

Asking @qxcv to take a quick look to sanity check high-level structure, and @ejnnr for a full review (although feel free to skim over things that seem like minor / boiler plate changes, there's a lot of those unfortunately).

Fixes #246; makes progress on #79

### CODE REVIEW

**Review by ejnnr — APPROVED** [review]
Thanks for this PR (and for using `RewardNet` in IRL, I think we can finally close #246 with this)! Inline comments are all just cosmetic.

In general, I think this is certainly an improvement, though I would have hoped we could share even more code (e.g. IRL and density still have their own implementations of `set_demonstrations`). But maybe that's just unrealistic, at least I don't see an obvious better abstraction.

There are some other refactors that would probably make sense (e.g. I think `DensityAlgorithm` could just expose a `RewardFn` and its `train_policy` and `test_policy` methods could be delegated to helper functions independent of the specific algorithm). But that's for another PR (I might do more refactoring myself in October when I work on preference comparisons again).

File: src/imitation/algorithms/adversarial.py [hunk 1]
```diff
@@ -40,13 +41,13 @@ class AdversarialTrainer(base.BaseImitationAlgorithm):
 
     def __init__(
```
Comment (qxcv) [inline src/imitation/algorithms/adversarial.py:42]: Nit: if we're going to mess around with argument order, should we make all of these arguments kwarg-only? That way we can do that without any risk of silently breaking existing code.
(the breakage here is not going to be silent anyway, which is good, but perhaps changing this now gives us more flexibility in the future)

Comment (AdamGleave) [reply src/imitation/algorithms/adversarial.py:42]: Good point, I generally prefer kwargs anyway for functions with many arguments, I'll make that mandatory everywhere.

File: src/imitation/algorithms/tabular_irl.py [hunk 2]
```diff
@@ -111,7 +117,11 @@ def mce_occupancy_measures(
             E = D[t - 1] * pi[t - 1, :, a]
             D[t, :] += E @ T[:, a, :]
 
-    return D, D.sum(axis=0)
+    if discount == 1.0:
+        Dcum = D.sum(axis=0)
+    else:
+        Dcum = np.polynomial.polynomial.polyval(discount, D)
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/tabular_irl.py:123]: Agreed, I've used this function (renamed to `discounted_sum` as `Dcum` isn't really returns).

File: src/imitation/algorithms/tabular_irl.py [hunk 3]
```diff
@@ -130,144 +140,178 @@ def squeeze_r(r_output: th.Tensor) -> th.Tensor:
...
-            layer_list.append(activation())
-            prev_h_size = h_size
-        layer_list.append(nn.Linear(prev_h_size, 1))
-        self.layers = nn.Sequential(*layer_list)
-
-    def forward(self, obs: th.Tensor) -> th.Tensor:
-        return th.squeeze(self.layers(obs), 1)
+        # use the same device and dtype as the rmodel parameters
+        a_param = next(self.reward_net.parameters())
+        device = a_param.device
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/tabular_irl.py:262]: Good idea, I've added a `dtype` property. @qxcv @scottemmons can you remember why we have a default for if the module has no parameters? When would it have no parameters -- I guess a constant zero reward network? I'm not sure what a default `dtype` even should be (I went with `float32`).

[the remark this answers is not in the document]
Comment (qxcv) [reply src/imitation/algorithms/tabular_irl.py]: Yeah, I think there is (or was) a constant reward network used for tests somewhere. This was to get around that.
(also generally I think it's a good idea not to assume that `nn.Module`s have any parameters, since there are many built-in ones that are parameter free, like `nn.ReLU`)

[the remark this answers is not in the document]
Comment (ejnnr) [reply src/imitation/algorithms/tabular_irl.py]: I think I added that default in the `RewardNet` refactor, at least it's not yet there in this commit: https://github.com/HumanCompatibleAI/imitation/blob/777085d16bf62d63020bc7d0f5d6a740728ff4ad/src/imitation/rewards/reward_nets.py#L185
My only motivation was that it shouldn't raise an error if a `RewardNet` doesn't have parameters for some reason, I didn't have a specific case in mind. But maybe I'm also misremembering and saw it somewhere else (SB3?) and just used it because of that

[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/tabular_irl.py]: You're right, it was added in https://github.com/HumanCompatibleAI/imitation/pull/311/files#diff-58b170ba20f1518bab467e4e002302c8a873e5e58b9319a4de15231c38766b3cR87 I think it's nice to have a default as a kind of defensive coding, but I'm going to not worry too much what the default is for now until we have a concrete use case. Mildly worried that this is going to cause silent failures down the line, but I can't think of any concrete case (if there are no parameters, shouldn't matter too much what you assume about their device/dtype...)

**Comment by AdamGleave** [comment]
Thanks both for the review.

> In general, I think this is certainly an improvement, though I would have hoped we could share even more code (e.g. IRL and density still have their own implementations of set_demonstrations). But maybe that's just unrealistic, at least I don't see an obvious better abstraction.

I'm also a bit unsatisfied by this. The algorithms are just quite different -- e.g. density needs a full set of transitions, while BC and GAIL/AIRL want a data loader yielding batches. I originally had data loader as the standard, but this loses timestep information that density needs for a non-stationary model. One could fix this by adding the timestep information into the batches, of course, but it felt a bit perverse to effectively go `Sequence[Trajectory]->DataLoader->Sequence[Trajectory]`.

> There are some other refactors that would probably make sense (e.g. I think DensityAlgorithm could just expose a RewardFn and its train_policy and test_policy methods could be delegated to helper functions independent of the specific algorithm). But that's for another PR (I might do more refactoring myself in October when I work on preference comparisons again).

Yeah I agree there's lots of room for improvement still. I'm still unsure what API we want to be exposing. Some of the algorithms only learn a policy (e.g. BC), others a policy and reward (AIRL, kind-of GAIL), others only learn a reward (MCE IRL, Density). One option is to always expose a policy (since can always go reward->policy), and optionally expose a RewardFn. Other option is to just make both of them optional, and push it onto the callee to train a policy if they want. I'm hoping this will become clearer when trying to unify the training scripts.

File: src/imitation/rewards/reward_nets.py [hunk 4]
```diff
@@ -113,6 +113,16 @@ def device(self) -> th.device:
...
 
+    @property
+    def dtype(self) -> th.dtype:
+        """Heuristic to determine dtype of module."""
+        try:
+            first_param = next(self.parameters())
+            return first_param.device
+        except StopIteration:
+            # if the model has no parameters, default to float32
+            return th.float32
```
Comment (qxcv) [inline src/imitation/rewards/reward_nets.py:124]: Suggestion: `torch.get_default_dtype()`
